capsules/extra/src/ethernet_tap.rs

```diff
@@ -0,0 +1,657 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Leon Schuermann <leon@is.currently.online> 2025.
+// Copyright Tock Contributors 2025.
+
+//! # Userspace application TAP-like network driver
+//!
+//! This capsule provides a userspace syscall driver designed to expose a raw
+//! IEEE 802.3 (Ethernet) compatible network interface (resembling a
+//! Linux/FreeBSD TAP virtual network device).
+//!
+//! The driver multiplexes the underlying network interfaces across all
+//! applications. Applications can allow a receive buffer to the kernel, and all
+//! incoming frames will be placed into all applications' receive buffers. Every
+//! application may transmit frames, which are sent on the underlying Ethernet
+//! link.
+//!
+//! Incoming frames are sent to all processes (effectively mimicking the
+//! behavior of a MacVTap device in Virtual Ethernet Port Aggregator (VEPA)
+//! mode). This means that all processes can communicate over the Ethernet link,
+//! but processes will generally not able to talk to each other, _unless_ the
+//! connected far-end Ethernet device supports hairpin mode and is able to send
+//! frames back on the link they came from. If processes should be able to talk
+//! to each other, we can instantiate a hairpinning bridge in the Tock kernel,
+//! or use a regular bridge and provide each process its own, dedicated tap
+//! driver.
+//!
+//! ## Interface description
+//!
+//! This driver can both deliver incoming IEEE 802.1 Ethernet frames from the
+//! backing [`EthernetAdapterDatapath`] interface to userspace applications, as
+//! well as accept frames from the application for transmission over the
+//! [`EthernetAdapterDatapath`] interface.
+//!
+//! The driver is built on top of the [`StreamingProcessSlice`] abstraction,
+//! which allows for lossless streaming of data from the kernel to a userspace
+//! application. It places incoming Ethernet frames into the
+//! [`StreamingProcessSlice`], prefixed by a driver-specific header per
+//! frame. This header contains each frames' length, allowing userspace to
+//! efficiently find frame delimitors without parsing their content.
+//!
+//! The per-frame header has the following format:
+//! - Bytes 0 to 1 (`u16`, native endian): flags
+//!   - Bit 0: receive timestamp valid
+//! - Bytes 2 to 3 (`u16`, native endian): frame length in bytes, excluding this header
+//! - Bytes 4 to 11 (`u64`, native endian): receive timestamp
+//! - Bytes 12 to (12 + frame length): frame contents, starting with the
+//!   Ethernet header, up to and exluding the Frame Check Sequence (FCS)
```

Comment:
The ordering here is unfortunate from an alignment standpoint. If the `u64` was first, you wouldn't have that issue ... unless the entire header isn't guaranteed to start at an 8-byte boundary either, which it might not be given that we're just putting packet-after-packet in the buffer.

Reply:
Right, exactly my thought. Neither side can rely on alignment, actually, as the app can allow the buffer at any address, and the headers (starting with the second frame) are placed in the buffer right after the previous frame, which may have any length.

Reply:
Makes sense